File: src/main/kotlin/com/store/models/Product.kt
package com.store.models

import javax.validation.constraints.NotBlank
import javax.validation.constraints.NotNull
import javax.validation.constraints.Min

data class Product(
    var id: Int? = null,
    
    @NotBlank(message = "Product name cannot be null or empty")
    var name: String,

    @NotNull(message = "Product type cannot be null")
    var type: Type,

    @Min(value = 1, message = "Inventory must be greater than zero")
    var inventory: Int,

    // @Min(value = 0.1, message = "Cost must be greater than zero")
    var cost: Double

) {
    enum class Type {
        book, food, gadget, other
    }
}

File: server/models/product.go
package models

type Product struct {
	Id        int    `json:"id"`
	Name      string `json:"name" binding:"required"`
	Type      string `json:"type" binding:"required"`
	Inventory int    `json:"inventory" binding:"required"`
	Cost      int    `json:"cost" binding:"required"`
}

var ProductType []string = []string{"book", "food", "gadget", "other"}

func IsValidProductType(product string) bool {
	if product == "" {
		return true
	}

	for _, n := range ProductType {
		if product == n {
			return true
		}
	}

	return false
}

type ProductId struct {
	Id int `json:"id"`
}
